File: runtime/jvm/src/lox/LoxError.java
package lox;

/**
 * A Lox++ runtime error. Uncaught, it aborts the program (see vm.cpp
 * runtimeError).
 */
public final class LoxError extends RuntimeException {
    /**
     * The Lox++ value being thrown; null for an internally-raised fault
     * (caught as an Error instance).
     */
    private final Object value;

    /** Construct a LoxError with a message string (fault, not throw). */
    public LoxError(String message) {
        super(message);
        this.value = null;
    }

    /** Construct a LoxError carrying a user-thrown Lox++ value. */
    public LoxError(String message, Object value) {
        super(message);
        this.value = value;
    }

    /**
     * True if this fault reaches a live Lox try/catch (and, on the
     * exceptional-exit path, a function's own pending defers — issue #319):
     * mirrors runtime/clr/src/LoxError.cs's Catchable property, and
     * src/vm.cpp's split between a fault raised through
     * tryCatchableError/raiseThrowableError (reaches handleThrow) and one
     * raised through RAISE_ERROR/plain runtimeError (never does). Unlike
     * {@link #getValue()}, this has no side effect, so callers that only
     * need the flag — not the wrapped value — can check it without
     * triggering getValue()'s own rethrow-if-uncatchable behavior.
     */
    public boolean isCatchable() {
        return value != null;
    }

    /**
     * Get the Lox++ value being thrown. Returns the original value if this
     * error wraps a user throw. If this is an internally-raised fault
     * (value == null), rethrow the exception instead of returning it,
     * making internally-raised faults uncatchable by Lox try/catch blocks.
     * This is the mechanism that distinguishes catchable errors (from user
     * throw or makeError() with an Error instance) from fatal system faults.
     */
    public Object getValue() {
        if (value == null) {
            throw this;
        }
        if (LoxClosure.isOverflowInFlight(this)) {
            // This is the fault LoxClosure is watching for (the
            // StackOverflowError it raised, or whatever replaced it — see
            // LoxClosure.s_overflowInFlight's own comment), and it is about
            // to reach a real catchBlock, so that unwind is over.
            LoxClosure.endStackOverflowUnwind();
        }
        return value;
    }
}

File: runtime/clr/src/LoxError.cs
using System;

namespace Lox;

/// <summary>
/// A Lox++ runtime error. Uncaught, it aborts the program (see vm.cpp runtimeError).
/// When constructed with a Value (for a try/catch scenario), the Value holds an Error
/// instance with message and kind fields per spec/03-types.md.
///
/// <see cref="Catchable"/> mirrors src/vm.cpp's own split between a fault
/// raised through <c>tryCatchableError</c>/<c>raiseThrowableError</c> (goes
/// through <c>handleThrow</c>, so a live Lox++ <c>try</c> can catch it) and
/// one raised through <c>RAISE_ERROR</c>/plain <c>runtimeError</c> (bypasses
/// <c>handleThrow</c> entirely, so no handler ever sees it, however many
/// live <c>try</c> blocks surround the fault site). The emitted CLR
/// <c>catch [LoxRuntime]Lox.LoxError</c> block (src/backend/clr_emitter.cpp)
/// reads this flag before running any handler body and rethrows when it is
/// false, so an uncatchable fault propagates past every live handler here
/// too, instead of being delivered with <see cref="Value"/> silently null.
/// </summary>
public sealed class LoxError : Exception {
    /// <summary>The Error instance being thrown (if this is a caught/catchable error).</summary>
    private object _value;
    public object Value { get { return _value; } }

    private bool _catchable;
    public bool Catchable { get { return _catchable; } }

    /// <summary>Construct from a message only (for uncatchable internal errors).</summary>
    public LoxError(string message) : base(message) {
        _value = null;
        _catchable = false;
    }

    /// <summary>
    /// Construct from a value (for a catchable runtime fault, built via
    /// <see cref="LoxRuntime.MakeError"/>, or for the exact value a Lox++
    /// <c>throw</c> statement passed — which reaches a live handler
    /// regardless of its own type, including <c>nil</c>).
    /// </summary>
    public LoxError(object value) : base(ExtractMessage(value)) {
        _value = value;
        _catchable = true;
    }

    private static string ExtractMessage(object value) {
        if (value is LoxInstance inst && inst.Fields.TryGetValue("message", out object msg)) {
            return msg as string ?? "Error";
        }
        return "Error";
    }
}
